### ISSUE TITLE
dots.mocr "filtered" pages: what is lost, and what to find out

### ISSUE THREAD

**Issue by grossir** [freelawproject/scanning#242]
# dots.mocr "filtered" pages: what is lost, and what to find out

Follow-up to #238 and PR #239.

## What a filtered page is

The worker asks dots.mocr for one JSON list per page: one object per
region with `bbox`, `category` and `text`. Upstream's
`post_process_output` calls `json.loads` on the answer. When the parse
fails, upstream does not retry and does not repair the JSON. It hands
the answer to `OutputCleaner.clean_model_output`, which strips the JSON
syntax and keeps the words, and returns `(cleaned_text, True)`. The
worker stores that as `filtered: true`, `cells: null`, `md: <text>`.

The model produced these answers in normal length (1171 to 1779 tokens
in the sample) on normal body pages. One wrong character in about 1500
tokens of structured output breaks the parse: an unescaped quotation
mark inside a `text` value, a missing comma, an unbalanced bracket.
Legal text is full of quotation marks and brackets. Nothing about the
input is wrong.

## How often

In the 136 shard results pulled for #238 (13159 pages): 17 filtered
pages, about one in 770. The sample is biased toward shards that also
had a looping page, so the corpus-wide rate is unknown. Until PR #239
a filtered page reached no log line and no database field, so the
only way to count them is to read every shard result object on S3.

## What is lost on a filtered page

- **The page number.** The reader (`page_numbers.py`) needs a cell in
  the head or foot band to place a number. A filtered page has no cell,
  so it gets `detected=None`, the sequence analysis raises
  `no_page_number`, and the number is interpolated. The number is in
  fact present in `md` as plain text (`104 Kan.`, `Mass. 1233`), but
  nothing reads it from there.
- **The layout.** Every `category` and every `bbox` of the page: the
  running head, the body columns, the footnotes, the pictures. Any
  future consumer of the layout (the smart glue over page edits, text
  alignment against the CourtListener opinion) has nothing for the
  page.
- **The structure of the text.** The cleaned `md` is the words in the
  order the model wrote them, with the JSON removed. Reading order
  per cell, cell boundaries and category labels are gone.
- **The raw answer, for every run before PR #239.** The cleaner
  consumed it and the worker stored only the words. So for the
  existing corpus nothing can say which character failed the parse,
  and no repair can be attempted offline. PR #239 keeps the answer in
  `raw` from now on.

A filtered page is a hole to the reader in exactly the way a failed
page is, and PR #239 now treats it so: `filtered_pages` in the worker
summary, an INFO line, `has_unread_pages` for the carry and the
backfill. What PR #239 does **not** do is fix the page: the only retry
is the thresholded render, which was designed for the show-through
loops, and its effect on a JSON slip on a normal page is unknown.

## What needs research

1. **Count them across the corpus.** A read-only script over every
   `jobs/analyze/dots_mocr/*.json` result object on S3, counting
   `filtered` pages per volume and per reporter. Gives the true rate
   and says whether some reporters or page kinds are worse.
2. **Classify the parse failures.** After the PR #239 image is live,
   read `raw` on new filtered pages and name the break: quote inside a
   value, comma, bracket, truncation, something else. A few dozen
   cases are enough to see the distribution.
3. **Measure the retry.** The INFO lines in Loki give, per shard, the
   filtered pages and the recovered pages. Read the share of filtered
   pages the thresholded render recovers. If it is low, the retry
   rung is the wrong tool for this class and should not be paid for
   it.
4. **Repair instead of re-read.** A JSON repair step (for example the
   `json_repair` package) between `json.loads` failing and the cleaner
   would recover the cells on most single-character slips at no GPU
   cost. Needs the classification from item 2 to say how many it
   would save, and a check that a repaired cell's `bbox` is still
   legal (`is_legal_bbox`). Worker-side, so it needs an image build.
5. **Read the page number from the text as a last resort.** The
   cleaned `md` starts with the running head. A reader that takes the
   first line of `md` when there is no cell would place a number with
   no geometry to check it against. Decide whether a number with no
   band is better than an interpolated one, given the rival numbers
   #228 found in the head band.
6. **Recover the existing corpus.** The 17 known pages, and whatever
   item 1 finds, have no `raw` and no `filtered_pages` on their rows.
   Options: a re-glue that reads `filtered` off the result objects and
   writes it into the rows, so the backfill can find them; or accept
   the manual page number in review 1 as the backstop for pages read
   before PR #239.
7. **Check upstream.** Whether a newer dots.ocr commit changed the
   cleaner or added a repair, and whether the pinned commit
   (`23f3e56`) is behind on this.

## Not in scope

The repetition loops of #238. Those are a different fault with a
targeted fix in PR #239.

**Comment by grossir** [comment]
One filtered page, read end to end. The answer was good. One character failed it.

The shard: `/scans/2726/glued/dots-mocr/r1/shards/9/` (scan 2726, run 1, shard 9, 92 pages). Its summary holds `filtered_pages: [44]`, `failed_pages: []` and `recovered_pages: []`.

## The fault

The `raw` of shard-local page 44 is a complete layout JSON array. It starts with `[{`, it ends with `}]`, and 91 of the 92 pages of the shard parse without a change. Page 44 stops at offset 3698, where the model wrote:

```
"text": "... her husband's death \"almost took [her] out[.]" and she felt like she \"couldn't make it without him[.]\" ..."
```

The model put a backslash in front of the first quotation mark of the quotation. It did not put one in front of the second. The JSON string closes early:

```
Expecting ',' delimiter: line 1 column 3701 (char 3700)
```

Upstream's `post_process_output` then discards the array and returns the cleaned text, so the handler writes `cells: None` and keeps the words in `md`.

## The cost

I added the one missing backslash. The array parses and gives 19 cells. Two of them are `Page-header`:

| bbox | text |
|---|---|
| `[276, 93, 426, 129]` | `878 N. C.` |
| `[519, 93, 1187, 125]` | `868 SOUTH EASTERN REPORTER, 2d SERIES` |

The printed number **878** is in the answer, at the outer left corner, which is where `page_numbers.py` ranks it first. The filter throws away all 19 cells, so the page offers no cell, the apply reads `detected=None`, and `_resolve_by_neighbours` or the interpolation has to cover a page the model read correctly. The 18 other cells (the text blocks, the headnote section headers) go with it. This is the `filtered_pages` hole #242 describes, and here it is one character wide.

## The ladder cannot answer it (#238)

The page carries `attempts: 2` and `errors: ["model output was not layout JSON", "model output was not layout JSON"]`. Rung 2 changes the render only (threshold 100). The fault is neither show-through nor a repetition loop: `completion_tokens` is 1619, against 2504 as the maximum of the shard and the cap of 6144. The model gave a full answer and made one mistake in the escape, so a second render of the same page makes the same class of mistake. The second rung spent 91 seconds and recovered nothing.

The carry rule sees the same thing from the other side: the hole is stable (`jobs.hole_is_stable`), so `reread_failed_pages` skips this volume for good. Only a changed worker recovers the page.

## The fix this page asks for

A repair pass in `handler._parse_page`, before the ladder climbs:

1. Parse `raw`. If the parse fails at a quotation mark, put a backslash in front of that quotation mark and parse again.
2. Repeat up to a small limit (this page needs one repair; stop at three, and accept the failure after that).
3. Keep `raw` as the model wrote it. Mark the page (`repaired_by: "json"`) and count the pages in the summary beside `recovered_pages`, so the next deploy shows how much the repair recovers.

Put the repair ahead of rung 2. The repair is microseconds and the rung is 91 seconds of GPU time.

Two limits to state. The repair reaches a lone quotation mark inside a string, which is the fault measured here; it does not reach a truncated array or a broken bbox. And a repaired page is still the model's own reading, so it needs no separate review.

Next: count the fault over the corpus. Every shard result object is kept, so a pass over the stored `filtered_pages` says how many of the holes are one quotation mark and how many need more.

**Comment by grossir** [comment]
We will need
- a fix on the dots.mocr image itself, to prevent paying for OCR when we just needed to fix the JSON
- the same fix to be reused on the glue for everything that already processed but has a saved "raw" value 
- the daemon should log a explicit WARNING for filtered pages, so we can catch more edge cases easily

**Comment by grossir** [comment]
A second instance, on another volume. The fault is the same class and a different shape, which changes the fix I proposed above.

The shard: `/scans/2702/glued/dots-mocr/r1/shards/6/` (scan 2702, run 1, shard 6, 98 pages). `filtered_pages: [32]`, `failed_pages: []`, `recovered_pages: []`. One page again, and one character again.

## The fault

The parser stops earlier than on scan 2726, and with another message:

```
Invalid \escape: line 1 column 1778 (char 1777)
```

The model wrote:

```
... at \"14.1 million\" to \"18.2 million adults\)))."
```

The page prints `"18.2 million adults")))."`, so the model had to write `\")))`. It wrote `\)))`. The backslash came out and the quotation mark it belongs to did not. Page 44 of scan 2726 failed the other way round: there the quotation mark came out and the backslash did not.

| | 2726 / shard 9 / page 44 | 2702 / shard 6 / page 32 |
|---|---|---|
| the model wrote | `out[.]"` | `adults\)))` |
| it had to write | `out[.]\"` | `adults\")))` |
| the parser says | `Expecting ',' delimiter` | `Invalid \escape` |
| lost | the backslash | the quotation mark |

## The measurement

I put the one missing quotation mark back. The array parses and gives 7 cells. Three are `Page-header`:

| bbox | text |
|---|---|
| `[667, 99, 966, 129]` | `BARNETT v. RAOUL` |
| `[611, 132, 1023, 159]` | `Cite as 756 F.Supp.3d 564 (S.D.Ill. 2024)` |
| `[1339, 101, 1397, 136]` | `621` |

The printed number **621** is there, alone in its cell at the outer right corner, with the two rivals of #228 (the case name, and page 564 of the `Cite as` line) nearer the middle. The geometric rank reads 621. The filter throws away all three cells, so the page offers none.

The rest of the shard is clean: 97 of the 98 pages parse with no change, and the longest answer is 1796 tokens against the cap of 6144. The page carries `attempts: 2` with the same error twice, so rung 2 again spent 120 seconds and recovered nothing. The fault is not the render.

## What this changes

**The trigger is now visible.** Both pages are dense with quotation marks: one quotes a witness, the other quotes an expert report inside three levels of parentheses. Both faults sit on a quotation mark inside `text`. The model handles a quotation mark inside a JSON string badly, and a page full of them raises the chance that one comes out wrong.

**The repair pass I described above is too narrow.** It adds a backslash in front of a lone quotation mark. That answers scan 2726 and does nothing for scan 2702. The pass needs one arm for each shape:

1. `Expecting ',' delimiter` at a quotation mark: put a backslash in front of that quotation mark.
2. `Invalid \escape`: put a quotation mark after the backslash.

Both arms restore what the page prints, and both move one character. Keep the limit small and keep `raw` as the model wrote it.

**This is the argument for constrained decoding.** Two shards, two instances, two shapes of one fault, and a repair pass grows an arm for every new shape a volume shows. `handler._vllm_inference` sends a plain chat completion: no `response_format` and no `guided_json`, so the JSON is a habit from training and nothing holds the model to it. A JSON schema on that call removes the whole class, because the decoder cannot emit a bare `"` inside a string and cannot emit `\)` at all. Two costs to measure first: the grammar masks the logits on every token, which slows generation, and a schema narrower than what the model really sends would turn a good page into a failure. Write the schema against the array the model sends, not the "single JSON object" upstream's prompt asks for.

The two fixes are not rivals. The repair recovers the shard results already in the bucket, which no new worker image reaches. Constrained decoding stops the next one.

**Comment by grossir** [comment]
A third instance. It repeats the first shape, so the fault class is now measured on three volumes and the two-arm repair answers all three.

The shard: `/scans/2665/glued/dots-mocr/r1/shards/2/` (scan 2665, run 1, shard 2, 100 pages). `filtered_pages: [27]`, `failed_pages: []`, `recovered_pages: []`. One page, one character, and 99 of the 100 pages parse with no change.

## The fault

```
Expecting ',' delimiter: line 1 column 4044 (char 4043)
```

The model wrote, inside one `text` value:

```
... filings were \"in effect, 'successive petitions for Post Relief Conviction[.] [sic]'" (Footnote omitted.) The court \"thoroughly reviewed ...
```

It put a backslash in front of the opening quotation mark and left the closing one bare, which is the shape of scan 2726. The JSON string ends at `[sic]'`, and the parser reads everything after it as syntax.

## The measurement

I added the one backslash. The array parses and gives 10 cells — `Text`, `Footnote`, and two `Page-header`:

| bbox | text |
|---|---|
| `[271, 115, 415, 151]` | `228 Ohio` |
| `[512, 115, 1169, 149]` | `203 NORTH EASTERN REPORTER, 3d SERIES` |

The printed number **228** is at the outer left corner and the rival 203 sits in the middle band, so the geometric rank of #228 reads 228. The filter throws away both cells and the eight others.

The page carries `attempts: 2` with the same error twice, so rung 2 spent 92 seconds and recovered nothing. `completion_tokens` is 1235 against 1798 as the maximum of the shard: no loop, and no truncation.

## The three instances

| | 2726 / s9 / p44 | 2702 / s6 / p32 | 2665 / s2 / p27 |
|---|---|---|---|
| the parser says | `Expecting ',' delimiter` | `Invalid \escape` | `Expecting ',' delimiter` |
| lost | the backslash | the quotation mark | the backslash |
| repairs needed | 1 | 1 | 1 |
| number recovered | 878 | 621 | 228 |
| the page quotes | a witness | a report inside three parentheses | a quotation inside a quotation |

Three volumes, three shards, one filtered page in each, one character in each, and the printed page number recovered every time. The shape count stays at two, so the two-arm repair answers all three: arm 1 for this page and for page 44, arm 2 for page 32.

The trigger holds as well. Every one of the three pages nests quotation marks, and this page nests `'...'` inside `"..."`. The model writes the escape correctly hundreds of times in each shard and fails once, on the page that asks for it most often.

Next: run the two-arm repair over every stored result object that holds a filtered page. The results are kept for good, so the pass costs no GPU time and it says what share of the corpus holes are one character.

**Comment by grossir** [comment]
A fourth instance, and a third shape.

The shard: `/scans/2705/glued/dots-mocr/r1/shards/4/` (scan 2705, run 1, shard 4, 99 pages). `filtered_pages: [72]`, `failed_pages: []`. 98 of the 99 pages parse with no change.

```
Extra data: line 1 column 4346 (char 4345)
```

The answer ends `... his territory either"}]"}]`. The model closed the string, the object and the array, then wrote the same three characters again. The array is complete at the first closer, so everything after it is extra. Note that this is the actual ending of the page, so the extraction is correct, what's wrong is the JSON formatting

## The measurement

I cut the answer at the first complete parse. It gives 8 cells, three of them `Page-header`:

| bbox | text |
|---|---|
| `[532, 110, 1139, 142]` | `HENRY v. NEW JERSEY TRANSIT CORP.` |
| `[662, 146, 1009, 170]` | `Cite as 210 N.E.3d 451 (N.Y. 2023)` |
| `[1264, 113, 1417, 148]` | `N. Y. 473` |

The printed number **473** is at the outer right corner, with the two rivals of #228 nearer the middle. The rank reads 473.

Nothing is lost. The last cell stops mid-sentence, and the printed page stops at the same word: the opinion continues on the next page, and the curly quotation mark it leaves open closes there. This is not the truncation of #238 either — `completion_tokens` is 1265, against 2187 as the maximum of the shard and the cap of 6144, so no rung raised `TruncatedOutput`.

## The repair, with three arms

| shape | seen on | the arm |
|---|---|---|
| a lone `"` inside a string | 2726 p44, 2665 p27 | put a backslash in front of it |
| `\)` instead of `\")` | 2702 p32 | put a quotation mark after the backslash |
| a doubled closer | 2705 p72 | cut at the first complete parse |

Four volumes, four filtered pages, four repairs of one edit each, and the printed page number recovered every time. Each arm gives the page back whole.
